common/src/main/java/dev/isxander/yacl3/gui/controllers/dropdown/AbstractDropdownControllerElement.java

```diff
@@ -0,0 +1,193 @@
+package dev.isxander.yacl3.gui.controllers.dropdown;
+
+import com.mojang.blaze3d.platform.InputConstants;
+import com.mojang.blaze3d.vertex.PoseStack;
+import dev.isxander.yacl3.api.utils.Dimension;
+import dev.isxander.yacl3.gui.YACLScreen;
+import dev.isxander.yacl3.gui.controllers.string.StringControllerElement;
+import dev.isxander.yacl3.gui.utils.GuiUtils;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.network.chat.Component;
+
+import java.util.List;
+
+public abstract class AbstractDropdownControllerElement<T, U> extends StringControllerElement {
+	public static final int MAX_SHOWN_NUMBER_OF_ITEMS = 7;
+	private final AbstractDropdownController<T> dropdownController;
+	protected boolean opened = false;
+	// Stores the current selection position. The item at this position in the dropdown list will be chosen as the
+	// accepted value when the element is closed.
+	protected int selected = 0;
+
+	public AbstractDropdownControllerElement(AbstractDropdownController<T> control, YACLScreen screen, Dimension<Integer> dim) {
+		super(control, screen, dim, false);
+		this.dropdownController = control;
+	}
+
+	public void open() {
+		opened = true;
+		selected = 0;
+	}
+
+	public void close() {
+		opened = false;
+		ensureValidValue();
+	}
+
+	protected int selected() {
+		return selected;
+	}
+
+	public void ensureValidValue() {
+		inputField = dropdownController.getValidValue(inputField, selected);
+	}
+
+	@Override
+	public boolean mouseClicked(double mouseX, double mouseY, int button) {
+		if (super.mouseClicked(mouseX, mouseY, button)) {
+			if (!opened) {
+				open();
+				doSelectAll();
+			}
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public void setFocused(boolean focused) {
+		if (focused) {
+			doSelectAll();
+			super.setFocused(true);
+		} else unfocus();
+	}
+
+	@Override
+	public void unfocus() {
+		close();
+		super.unfocus();
+	}
+
+	@Override
+	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
+		if (!inputFieldFocused)
+			return false;
+		if (opened) {
+			switch (keyCode) {
+				case InputConstants.KEY_DOWN -> {
+					int len = getDropdownLength();
+					selected = Math.min(selected + 1, len - 1);
+					return true;
+				}
+				case InputConstants.KEY_UP -> {
+					selected = Math.max(selected - 1, 0);
+					return true;
```

Comment:
Could tab be implemented here to cycle through as well?

Reply:
I originally did not implement tab support because that allows you to "tab-complete" the currently selected option by unfocusing the field. After playing around with both options a bit, I'm not sure which I prefer, so I'm keeping tab-to-navigate instead.